File: examples/metadata-schema/typescript/schema_validation.ts
/**
 * Validate metadata with Zod before writing to Chroma.
 *
 * Shows a full roundtrip:
 * 1) validate metadata in app layer
 * 2) insert into Chroma
 * 3) read back and validate again
 * 4) run a filtered query and parse top metadata result
 *
 * Requires a running Chroma server:
 *   docker run --rm -p 8000:8000 chromadb/chroma:1.5.1
 */

import { ChromaClient } from "chromadb";
import { z } from "zod";

const recordMetadataSchema = z
  .object({
    tenant_id: z.string().min(1),
    doc_type: z.enum(["policy", "faq", "runbook"]),
    published: z.boolean(),
    priority: z.number().int().min(1).max(5),
    quality_score: z.number().min(0).max(1),
  })
  .strict();

type RecordMetadata = z.infer<typeof recordMetadataSchema>;

type RawRecord = {
  id: string;
  document: string;
  embedding: number[];
  metadata: unknown;
};

const rawRecords: RawRecord[] = [
  {
    id: "doc-1",
    document: "Password reset policy for employees.",
    embedding: [0.1, 0.2, 0.3],
    metadata: {
      tenant_id: "acme",
      doc_type: "policy",
      published: true,
      priority: 2,
      quality_score: 0.91,
    },
  },
  {
    id: "doc-2",
    document: "FAQ for account lockouts.",
    embedding: [0.2, 0.3, 0.4],
    metadata: {
      tenant_id: "acme",
      doc_type: "faq",
      published: true,
      priority: 0,
      quality_score: 0.7,
    },
  },
  {
    id: "doc-3",
    document: "Runbook for SSO outage response.",
    embedding: [0.3, 0.4, 0.5],
    metadata: {
      tenant_id: "",
      doc_type: "runbook",
      published: false,
      priority: 1,
      quality_score: 0.88,
    },
  },
];

const client = new ChromaClient({ host: "localhost", port: 8000 });
const collectionName = "metadata_schema_typescript";

try {
  await client.deleteCollection({ name: collectionName });
} catch {
  // ignore
}

const collection = await client.createCollection({ name: collectionName });

const insertValidatedRecord = async (record: RawRecord): Promise<RecordMetadata | null> => {
  const validated = recordMetadataSchema.safeParse(record.metadata);
  if (!validated.success) {
    console.log(`rejected ${record.id}: ${validated.error.message}`);
    return null;
  }

  await collection.add({
    ids: [record.id],
    documents: [record.document],
    embeddings: [record.embedding],
    metadatas: [validated.data],
  });
  console.log(`inserted ${record.id}`);

  return validated.data;
};

const readValidatedRecord = async (id: string): Promise<RecordMetadata> => {
  const result = await collection.get({
    ids: [id],
    include: ["documents", "metadatas"],
  });

  const metadata = result.metadatas?.[0];
  if (!metadata) {
    throw new Error(`no metadata returned for ${id}`);
  }

  const parsed = recordMetadataSchema.parse(metadata);
  console.log(`read ${id} -> ${JSON.stringify(parsed)}`);
  return parsed;
};

const insertedIds: string[] = [];

for (const record of rawRecords) {
  const inserted = await insertValidatedRecord(record);
  if (inserted === null) {
    continue;
  }

  insertedIds.push(record.id);
  await readValidatedRecord(record.id);
}

if (insertedIds.length === 0) {
  throw new Error("no records passed schema validation");
}

const queryResults = await collection.query({
  queryEmbeddings: [[0.1, 0.2, 0.3]],
  nResults: 3,
  where: { tenant_id: "acme" },
  include: ["documents", "metadatas", "distances"],
});
console.log(`query ids: ${JSON.stringify(queryResults.ids)}`);

const topMeta = queryResults.metadatas?.[0]?.[0];
if (topMeta) {
  const parsed = recordMetadataSchema.parse(topMeta);
  console.log(`query top metadata (typed): ${JSON.stringify(parsed)}`);
}

File: examples/metadata-schema/python/schema_validation.py
"""Validate metadata with Pydantic before writing to Chroma.

Shows a full roundtrip:
1) validate metadata in app layer
2) insert into Chroma
3) read back and validate again
4) run a filtered query and parse top metadata result

Requires a running Chroma server, for example:
    docker run --rm -p 8000:8000 chromadb/chroma:1.5.1
"""

from typing import Any, Literal

import chromadb
from pydantic import BaseModel, ConfigDict, Field, ValidationError


class RecordMetadata(BaseModel):
    """Application-level metadata schema contract."""

    model_config = ConfigDict(extra="forbid")

    tenant_id: str = Field(min_length=1)
    doc_type: Literal["policy", "faq", "runbook"]
    published: bool
    priority: int = Field(ge=1, le=5)
    quality_score: float = Field(ge=0.0, le=1.0)


class RawRecord(BaseModel):
    id: str
    document: str
    embedding: list[float]
    metadata: dict[str, Any]


def insert_validated_record(collection: Any, record: RawRecord) -> RecordMetadata | None:
    """Validate one record and insert into Chroma if valid."""

    try:
        validated = RecordMetadata.model_validate(record.metadata)
    except ValidationError as exc:
        print(f"rejected {record.id}: {exc.errors()}")
        return None

    collection.add(
        ids=[record.id],
        documents=[record.document],
        embeddings=[record.embedding],
        metadatas=[validated.model_dump()],
    )
    print(f"inserted {record.id}")
    return validated


def read_validated_record(collection: Any, record_id: str) -> RecordMetadata:
    """Read one record by id and parse metadata back into typed schema."""

    result = collection.get(ids=[record_id], include=["documents", "metadatas"])
    if not result["metadatas"] or not result["metadatas"][0]:
        raise ValueError(f"no metadata returned for {record_id}")

    parsed = RecordMetadata.model_validate(result["metadatas"][0])
    print(f"read {record_id} -> {parsed.model_dump()}")
    return parsed


def main() -> None:
    client = chromadb.HttpClient(host="localhost", port=8000)
    collection_name = "metadata_schema_python"

    try:
        client.delete_collection(name=collection_name)
    except Exception:
        pass

    collection = client.create_collection(name=collection_name)

    raw_records = [
        RawRecord(
            id="doc-1",
            document="Password reset policy for employees.",
            embedding=[0.1, 0.2, 0.3],
            metadata={
                "tenant_id": "acme",
                "doc_type": "policy",
                "published": True,
                "priority": 2,
                "quality_score": 0.91,
            },
        ),
        RawRecord(
            id="doc-2",
            document="FAQ for account lockouts.",
            embedding=[0.2, 0.3, 0.4],
            metadata={
                "tenant_id": "acme",
                "doc_type": "faq",
                "published": True,
                "priority": 6,
                "quality_score": 0.7,
            },
        ),
        RawRecord(
            id="doc-3",
            document="Runbook for SSO outage response.",
            embedding=[0.3, 0.4, 0.5],
            metadata={
                "tenant_id": "",
                "doc_type": "runbook",
                "published": False,
                "priority": 1,
                "quality_score": 0.88,
            },
        ),
    ]

    inserted_ids: list[str] = []

    for record in raw_records:
        validated = insert_validated_record(collection, record)
        if validated is None:
            continue

        inserted_ids.append(record.id)
        read_validated_record(collection, record.id)

    if not inserted_ids:
        raise RuntimeError("no records passed schema validation")

    query_results = collection.query(
        query_embeddings=[[0.1, 0.2, 0.3]],
        n_results=3,
        where={"tenant_id": "acme"},
        include=["documents", "metadatas", "distances"],
    )
    print(f"query ids: {query_results['ids']}")

    top_meta = (
        query_results.get("metadatas", [[]])[0][0]
        if query_results.get("metadatas") and query_results["metadatas"][0]
        else None
    )
    if top_meta is not None:
        parsed = RecordMetadata.model_validate(top_meta)
        print(f"query top metadata (typed): {parsed.model_dump()}")


if __name__ == "__main__":
    main()
